blogs/WrappedYieldAwaitPattern.md

```diff
@@ -0,0 +1,103 @@
+
+I called this the wrapped yield/<strong>await</strong> pattern, so async/await must come into this too, right? <a href="http://msdn.microsoft.com/en-us/library/hh191443.aspx">Async</a> methods are similar to yield iterators in that they are rewritten by the compiler to behave quite differently from normal control flow. Unlike yield, though, async methods are NOT lazy: when an async method is called, it executes synchronously until it hits an await. What's the issue, then? Let's imagine we have a method SkipLinesAsync(int) which is supposed to skip past lines in a <a href="http://msdn.microsoft.com/en-us/library/system.io.textreader%28v=vs.110%29.aspx">TextReader</a>. We could implement this as follows:
+
+<pre>
+public static async Task SkipLinesAsync(this TextReader reader, int linesToSkip)
+{
+	Console.WriteLine("validating arguments...");
+	if (reader == null) { throw new ArgumentNullException("reader"); }
+	if (linesToSkip < 0) { throw new ArgumentOutOfRangeException("linesToSkip"); }	
+	
+	for (var i = 0; i < linesToSkip; ++i)
+	{
+		var line = await reader.ReadLineAsync().ConfigureAwait(false);
+		if (line == null) { break; }
+	}
+}
+</pre>
+
+What happens if we run this?
+
+<pre>
+var reader = new StringReader("abc");
+
+// this line succeeds!
+var task = reader.SkipLinesAsync(-1); 
+// "validating arguments..." prints here. Still no failure
+Console.WriteLine("----");
+
+// finally, this line throws AggregateException (inner exception is ArgumentOutOfRangeException)
+task.Wait();
+</pre>
+
+Again, this is not very desirable: the exception is only reported when the task is observed (which could happen much later or even never for fire-and-forget tasks). The issue here is that everything inside an async method is considered part of the "Task" returned by that method. Thus, even though the failure happens synchronously, it is caught by the task system and stored in the Task itself (to be rethrown on Wait()). As with yield, we can fix this using a wrapper:
+
+<pre>
+public static Task SkipLinesAsync(this TextReader reader, int linesToSkip)
+{
+	if (reader == null) { throw new ArgumentNullException("reader"); }
+	if (linesToSkip < 0) { throw new ArgumentOutOfRangeException("linesToSkip"); }	
+	
+	return reader.SkipLinesInternalAsync(linesToSkip);
+}
+
+private static async Task SkipLinesInternalAsync(this TextReader reader, int linesToSkip)
+{
+	for (var i = 0; i < linesToSkip; ++i)
+	{
+		var line = await reader.ReadLineAsync().ConfigureAwait(false);
+		if (line == null) { break; }
+	}
+}
+</pre>
+
+Because the outer method is no longer async, it executes the validation outside of the context of the async Task, so get the synchronous, unwrapped validation errors we'd like.
```

Comment:
I added a section at the end addressing this. Thinking about it further, I think the behavior of the language is actually the lesser evil.